tempesta_fw/t/unit/test_sched_hash.c

```diff
@@ -43,96 +43,92 @@
+	"GET / HTTP/1.1\r\nhost:host3\r\n\r\n",
+	"GET / HTTP/1.1\r\nhost:host4\r\n\r\n",
 };
-size_t req_strs_size = sizeof(req_strs) / sizeof(req_strs[0]);
 
-TEST(tfw_sched_hash, sg_empty)
+static TfwMsg * sched_hash_get_scheduler_arg (size_t connection_type);
+
+static void
+sched_hash_free_scheduler_arg (TfwMsg * msg)
 {
-	int i, j;
+	test_req_free((TfwHttpReq *)msg);
+}
 
-	TfwSrvGroup *sg = test_create_sg("test", "hash");
+static struct TestSchedHelper sched_helper_hash = {
+	.scheduler = "hash",
+	.connections_types = sizeof(req_strs) / sizeof(req_strs[0]),
+	.get_scheduler_arg = &sched_hash_get_scheduler_arg,
+	.free_scheduler_arg = &sched_hash_free_scheduler_arg,
+};
 
-	for (i = 0; i < req_strs_size; ++i) {
-		for (j = 0; j < 3; ++j) {
-			TfwConnection *conn;
-			TfwHttpReq *req = test_req_alloc(strlen(req_strs[i]));
+static TfwMsg *
+sched_hash_get_scheduler_arg (size_t connection_type)
+{
+	TfwHttpReq *req = NULL;
 
-			tfw_http_parse_req(req, req_strs[i],
-					   strlen(req_strs[i]));
+	BUG_ON(connection_type >= sched_helper_hash.connections_types);
 
-			conn = sg->sched->sched_srv((TfwMsg *)req, sg);
-			EXPECT_TRUE(conn == NULL);
+	req = test_req_alloc(strlen(req_strs[connection_type]));
+	tfw_http_parse_req(req, (unsigned char *) req_strs[connection_type],
+					   strlen(req_strs[connection_type]));
 
-			test_req_free(req);
-		}
-	}
+	return (TfwMsg *) req;
+}
 
-	test_sg_release_all();
+TEST(tfw_sched_hash, sg_empty)
+{
+	test_sched_generic_empty_sg(&sched_helper_hash);
 }
 
 TEST(tfw_sched_hash, one_srv_in_sg_and_zero_conn)
 {
-	int i, j;
-
-	TfwSrvGroup *sg = test_create_sg("test", "hash");
-	test_create_srv("127.0.0.1", sg);
-
-	for (i = 0; i < req_strs_size; ++i) {
-		for (j = 0; j < 3; ++j) {
-			TfwConnection *conn;
-			TfwHttpReq *req = test_req_alloc(strlen(req_strs[i]));
-
-			tfw_http_parse_req(req, req_strs[i], strlen(req_strs[i]));
-
-			conn = sg->sched->sched_srv((TfwMsg *)req, sg);
-			EXPECT_TRUE(conn == NULL);
-
-			test_req_free(req);
-		}
-	}
-
-	test_sg_release_all();
+	test_sched_generic_one_srv_in_sg_zero_conn(&sched_helper_hash);
 }
 
+/*
+ * This unit test is implementation aware and checks more than just interface.
+ * Note, that it is very similar to other tests (one_srv_in_sg_and_max_conn and
+ * max_srv_in_sg_and_max_conn) for round-robin and hash schedullers. So if test
+ * structure is changed, other mentioned in above tests should be also be updated
+ */
 TEST(tfw_sched_hash, one_srv_in_sg_and_max_conn)
 {
-	int i, j;
+	size_t i, j;
 
-	TfwSrvGroup *sg = test_create_sg("test", "hash");
+	TfwSrvGroup *sg = test_create_sg("test", sched_helper_hash.scheduler);
 	TfwServer *srv = test_create_srv("127.0.0.1", sg);
 
 	for (i = 0; i < TFW_SRV_MAX_CONN; ++i) {
 		TfwSrvConnection *sconn = test_create_conn((TfwPeer *)srv);
 		sg->sched->add_conn(sg, srv, &sconn->conn);
+		/* connection reference count must be >0
+		 * otherwise connection will be scipped in scheduler
+		 */
```

Comment:
The first line of a multi-line comment should contain just a single /*. Then please start the sentence with a capital character. This applies to all comments within functions in this patch set.
`scipped` -> `skipped`.